### PR TITLE
feat: labeling all resources created by a zarf package

### LINKED ISSUE #3625: Add zarf label to all resources created by a package deploy
### Clear labeling of Zarf created resources
I want to be able to query k8s api and get a list of resources back that were created by deploying a zarf package. Something like `zarf-created-by=<package-name>` could be added to the resource labels.

### Describe the behavior you'd like
1. 
- **Given** a Zarf package
- **When** It is deployed
- **Then** All Kubernetes resources that are created by the deployment are labeled with the package name

2. 
- **Given** a Zarf package is deployed
- **When** `kubectl get <resource> -n <namespace> -l zarf-created-by=<pacakge-name>` is executed
- **Then** all resources created in the requested namespace by the specified package are returned

### PR DESCRIPTION
**Pull request by sgettys**
## Description

Adds labels to all resources created by a zarf package:
```
zarf.dev/package: <package name>
zarf.dev/namespace-override: <namespace if provided>
```

This will allow for easy filtering of resources associated with a zarf package using label selectors.

## Related Issue

Fixes #3625 

## Checklist before merging

- [x] Test, docs, adr added or updated as needed
- [x] [Contributor Guide Steps](https://github.com/zarf-dev/zarf/blob/main/CONTRIBUTING.md#developer-workflow) followed

### CODE REVIEW

File: src/api/v1alpha1/package.go [hunk 1]
```diff
@@ -47,6 +47,10 @@ const (
 	ZarfPackageConfig ZarfPackageKind = "ZarfPackageConfig"
 	// APIVersion the api version of this package.
 	APIVersion string = "zarf.dev/v1alpha1"
+	// PackageLabel is the label used to identify this package.
```
Comment (Racer159) [inline src/api/v1alpha1/package.go]:
```suggestion
	// PackageLabel is the label used to identify the owning package.
```
Just a suggestion to align with K8s wording https://kubernetes.io/docs/concepts/overview/working-with-objects/owners-dependents/

File: src/internal/packager/helm/chart.go [hunk 2]
```diff
@@ -54,6 +54,10 @@ type InstallUpgradeOptions struct {
 	Timeout time.Duration
 	// Retries for the helm install/upgrade
 	Retries int
+	// Pkg is the zarf package being installed
+	Pkg *v1alpha1.ZarfPackage
+	// NamespaceOverride is the namespace to use for the chart
```
Comment (Racer159) [inline src/internal/packager/helm/chart.go]:
```suggestion
	// NamespaceOverride is the namespace override to use for the chart
```
zarfChart.namespace is the namespace for the chart (which would still have a value even if the namespace override was empty)

File: src/internal/packager/helm/chart.go [hunk 3]
```diff
@@ -80,7 +84,7 @@ func InstallOrUpgradeChart(ctx context.Context, zarfChart v1alpha1.ZarfChart, ch
 		return nil, "", fmt.Errorf("unable to initialize the K8s client: %w", err)
 	}
 
-	postRender, err := newRenderer(ctx, zarfChart, opts.AdoptExistingResources, opts.Cluster, opts.AirgapMode, opts.State, actionConfig, opts.VariableConfig)
+	postRender, err := newRenderer(ctx, zarfChart, opts.AdoptExistingResources, opts.Cluster, opts.AirgapMode, opts.State, actionConfig, opts.VariableConfig, opts.Pkg, opts.NamespaceOverride)
```
Comment (Racer159) [inline src/internal/packager/helm/chart.go]: so much of `opts` is passed into `newRenderer` does it make sense to pass in `opts` as a whole?

File: src/internal/packager/helm/post-render.go [hunk 4]
```diff
@@ -117,6 +121,19 @@ func (r *renderer) adoptAndUpdateNamespaces(ctx context.Context) error {
 			}
 		}
 
+		// Add the package label to the namespace
+		if r.pkg != nil {
```
Comment (Racer159) [inline src/internal/packager/helm/post-render.go]: Did you get past the issue here with init package labeling?

File: src/test/e2e/26_simple_packages_test.go [hunk 5]
```diff
@@ -72,10 +75,60 @@ func TestManifests(t *testing.T) {
 
 	path := filepath.Join(tmpdir, fmt.Sprintf("zarf-package-manifests-%s-0.0.1.tar.zst", e2e.Arch))
 
-	// Deploy the package
-	stdOut, stdErr, err = e2e.Zarf(t, "package", "deploy", path, "--confirm")
+	// Deploy the package. Need to adopt existing resources to make sure that the label update for the package name
+	// applies to any previously created namespaces
+	stdOut, stdErr, err = e2e.Zarf(t, "package", "deploy", path, "--confirm", "--adopt-existing-resources")
+	require.NoError(t, err, stdOut, stdErr)
```
Comment (Racer159) [inline src/test/e2e/26_simple_packages_test.go]: we should change the behavior to add the label to namespaces without this flag IMO

File: src/internal/packager/helm/post-render.go (hunk 4 above)
Comment (sgettys) [reply src/internal/packager/helm/post-render.go]: Let me test that out again. I think we should assume that the renderer can only run when pkg is NOT nil?

Comment (Racer159) [reply src/internal/packager/helm/post-render.go]: we should though IMO we could cover this with tests - nil checks aren't bad (they help prevent outright crashes) but we should ensure that the behavior works as expected when the correct inputs are given.

Comment (sgettys) [reply src/internal/packager/helm/post-render.go]: The post renderer runs as part of `UpdateZarfRegistryValues` and `UpdateZarfAgentValues` which run outside the context of a package

File: src/internal/packager/helm/post-render.go [hunk 6]
```diff
@@ -111,19 +114,26 @@ func (r *renderer) adoptAndUpdateNamespaces(ctx context.Context) error {
...
+		// If the namespace doesn't exist then create it. If it does exist and is already managed by Zarf then update the labels with
+		// the new package and namespace override labels.
 		if !existingNamespace {
 			// This is a new namespace, add it
 			_, err := c.Clientset.CoreV1().Namespaces().Create(ctx, namespace, metav1.CreateOptions{})
 			if err != nil {
 				return fmt.Errorf("unable to create the missing namespace %s", name)
 			}
-		} else if r.adoptExistingResources {
+		} else if zarfManaged { // If the namespace is already managed by Zarf then update the labels with the new package and namespace override labels.
```
Comment (Racer159) [inline src/internal/packager/helm/post-render.go]:
```suggestion
		} else if zarfManaged || r.adoptExistingResources { // If the namespace is already managed by Zarf then update the labels with the new package and namespace override labels.
```
the adopt should still adopt everything

File: src/api/v1alpha1/package.go [hunk 7]
```diff
@@ -47,6 +47,10 @@ const (
 	ZarfPackageConfig ZarfPackageKind = "ZarfPackageConfig"
 	// APIVersion the api version of this package.
 	APIVersion string = "zarf.dev/v1alpha1"
+	// PackageLabel is the label used to identify the owning of package.
+	PackageLabel string = "zarf.dev/package"
+	// NamespaceOverrideLabel is the label used to identify the namespace override.
+	NamespaceOverrideLabel string = "zarf.dev/namespace-override"
```
Comment (AustinAbro321) [inline src/api/v1alpha1/package.go]: A better place to put these constants would be in this block https://github.com/zarf-dev/zarf/blob/main/src/pkg/cluster/cluster.go#L31-L38

File: src/internal/packager/helm/post-render.go [hunk 8]
```diff
@@ -40,32 +40,38 @@ type renderer struct {
...
-	if c == nil {
+func newRenderer(ctx context.Context, chart v1alpha1.ZarfChart, actionConfig *action.Configuration, opts InstallUpgradeOptions) (*renderer, error) {
+	if opts.Cluster == nil {
 		return nil, fmt.Errorf("cluster required to run post renderer")
 	}
 	if actionConfig == nil {
 		return nil, fmt.Errorf("action configuration required to run post renderer")
 	}
-	if variableConfig == nil {
+	if opts.VariableConfig == nil {
```
Comment (AustinAbro321) [inline src/internal/packager/helm/post-render.go]: I'd rather not have this function take opts. While it's semantically nicer, generally opts shouldn't be required. Also one of the goals with opts is to avoid breaking changes in function signatures, but since this is private we don't have to worry about that.

File: src/internal/packager/helm/post-render.go [hunk 9]
```diff
@@ -111,19 +117,26 @@ func (r *renderer) adoptAndUpdateNamespaces(ctx context.Context) error {
...
 			if serverNamespace.Name == name {
 				existingNamespace = true
+				if serverNamespace.Labels[state.ZarfManagedByLabel] == "zarf" {
+					zarfManaged = true
+				}
+				break
 			}
 		}
-
+		namespace.Labels = r.setPackageLabels(namespace.Labels)
```
Comment (AustinAbro321) [inline src/internal/packager/helm/post-render.go]: One implication that I'm thinking about here is that if two packages deploy resources to the same namespace then whichever deploy was more recent will take over and the label will change. Also, Zarf and Helm do not delete namespaces as well so these labels will stick around after the packages are removed from the cluster. Helm for instance doesn't add the `app.kubernetes.io/name` label to namespaces it creates. Still, these aren't necessarily a deal breaker to adding labels to namespaces as these side effects are relatively inconsequential, but I would like to evaluate it.

Could you provide more info about your use case @sgettys @Racer159. Is having this label on a namespace useful?

File: src/test/e2e/26_simple_packages_test.go [hunk 10]
```diff
@@ -76,6 +79,55 @@ func TestManifests(t *testing.T) {
...
+
+	deploymentList := &appsv1.DeploymentList{}
+	err = json.Unmarshal([]byte(stdOut), deploymentList)
+	require.NoError(t, err)
+	require.Len(t, deploymentList.Items, 3, "expected 3 deployments")
+
+	// Wait for all the deployments to be ready
+	stdOut, stdErr, err = e2e.Kubectl(t, "wait", "deployment", "-l", "zarf.dev/package=manifests", "--for=condition=Available", "--all-namespaces", "--timeout=1m")
+	require.NoError(t, err, stdOut, stdErr)
```
Comment (AustinAbro321) [inline src/test/e2e/26_simple_packages_test.go]: So this is a bit interesting, initially I was going to say that this is unnecessary since Zarf waits for resources to be ready by default, but it actually is necessary because the podinfo deployment has an hpa which _usually_ doesn't start scaling until after the deployment is already considered ready by Zarf. Probably it would be fine, but I don't want the hpa to take too long and not get caught by this wait, or make a third pod and fail the test. I think it's simpler to just test what we want in the dos-games test above.

File: src/pkg/packager/deploy.go [hunk 11]
```diff
@@ -290,7 +291,7 @@ func (d *deployer) deployInitComponent(ctx context.Context, pkgLayout *layout.Pa
 
 	// Before deploying the seed registry, start the injector
 	if isSeedRegistry {
-		err := d.c.StartInjection(ctx, pkgLayout.DirPath(), pkgLayout.GetImageDirPath(), component.Images, d.s.RegistryInfo.NodePort)
+		err := d.c.StartInjection(ctx, pkgLayout.DirPath(), pkgLayout.GetImageDirPath(), component.Images, d.s.RegistryInfo.NodePort, &pkgLayout.Pkg)
```
Comment (AustinAbro321) [inline src/pkg/packager/deploy.go]: I'd lean towards skipping the labeling of the injector. I can see the argument either way on if it'd be accurate, on one hand there is an injector component in the init package, on the other hand these resources are created by Zarf itself. Given that the injector is running for ~15 seconds, I'd rather air on keeping our code simpler unless there is a use case for it.

**Review by AustinAbro321 — CHANGES_REQUESTED** [review]
Thanks for making this change, I think it will be a positive for the project. Added a few nits and questions around scope

File: src/internal/packager/helm/post-render.go (hunk 9 above)
Comment (Racer159) [reply src/internal/packager/helm/post-render.go]: We don't actually need the namespaces - we were just trying to be complete here - I would be fine with just not labeling namespaces actually - the need we had was for tracking Pods / Pod Owner resources.

File: src/pkg/packager/deploy.go (hunk 11 above)
Comment (Racer159) [reply src/pkg/packager/deploy.go]: I asked for this to be included to more accurately see the pods in a package - the injector is normally only around for a little bit _but_ there are cases where Zarf init fails at the injector step and leaves the injector running - seeing the injector pod come back in a label selection query (i.e. if you have controlled K8s API access) can be a signal that this was the stage that Zarf init failed on.

File: src/internal/packager/helm/post-render.go (hunk 8 above)
Comment (sgettys) [reply src/internal/packager/helm/post-render.go]: I initially had it that way then went to opts because of the signature size and the fact that it was basically all of the "ops". I'm fine to revert back to explicit fields though

File: src/internal/packager/helm/post-render.go (hunk 9 above)
Comment (sgettys) [reply src/internal/packager/helm/post-render.go]: My thought around this was this label would be useful to at least denote the latest zarf package to operate against that namespace assuming multiple packages operate against it. We also discussed having a list of packages and appending instead of replacing but that could be confusing especially since you'd have to manage the label removal on package remove. In general I'm a fan of "labeling everything" including namespaces and this will at the bare minimum be a highly visible way to know if zarf has operated against a namespace or not

File: src/pkg/packager/deploy.go (hunk 11 above)
Comment (sgettys) [reply src/pkg/packager/deploy.go]: We discussed this and came to the decision to label even the injector just in case something failed during init. The philosophy of "zarf labels resources it creates" is what drove this in combination with debuggability on init failures

File: src/test/e2e/26_simple_packages_test.go (hunk 10 above)
Comment (sgettys) [reply src/test/e2e/26_simple_packages_test.go]: I do at least like selecting multiple resources from the same package and since "package pods" is really what I'm getting at with this whole labeling thing IMO this is a good testcase. I can make it less restrictive on number of pods though to account for hpa!

File: src/internal/packager/helm/post-render.go (hunk 9 above)
Comment (AustinAbro321) [reply src/internal/packager/helm/post-render.go]: I'm still somewhat torn, I could see it being useful on occasion. While it'd be rare, the argument against adding the label to namespaces is to avoid a user doing something such as making a script where they grab the namespaces deployed by their package, but suddenly having that script break when another package is deployed to that namespace. I agree that adding and removing different packages from the label is not worth the complexity. 

I'm curious what the other @zarf-dev/maintainers think.

File: src/pkg/packager/deploy.go (hunk 11 above)
Comment (AustinAbro321) [reply src/pkg/packager/deploy.go]: Fair enough

File: src/internal/packager/helm/post-render.go (hunk 9 above)
Comment (sgettys) [reply src/internal/packager/helm/post-render.go]: Now that we have this functionality available it will be very easy to add this in for namespaces down the road if the use cases start popping up. I'll remove for now

Comment (Racer159) [reply src/internal/packager/helm/post-render.go]: My 2 cents are that it is easy to add back if we need to in the future and can be left out as @sgettys mentioned - I do like completeness but if there are unexpected side effects that folks need to take into account it is worth holding off until we see more usecases - a list would be more complete but has its own side effects (what happens when a package is removed and the namespace is owned by two packages?).

File: src/test/e2e/26_simple_packages_test.go (hunk 10 above)
Comment (AustinAbro321) [reply src/test/e2e/26_simple_packages_test.go]: Sounds good, but you should check for one or more replicas. A flake happened here because the check occurred before the hpa bumped up the replica count to two. https://github.com/zarf-dev/zarf/actions/runs/18012570063/job/51253428537?pr=4194

Comment (sgettys) [reply src/test/e2e/26_simple_packages_test.go]: Yeah I added that. Getting ready to push

File: src/internal/packager/helm/zarf.go [hunk 12]
```diff
@@ -28,6 +28,11 @@ import (
 
 // UpdateZarfRegistryValues updates the Zarf registry deployment with the new state values
 func UpdateZarfRegistryValues(ctx context.Context, opts InstallUpgradeOptions) error {
+	pkg, err := opts.Cluster.GetDeployedPackage(ctx, "init")
```
Comment (AustinAbro321) [inline src/internal/packager/helm/zarf.go]: To make this work with custom named init packages we could use cluster.GetDeployedZarfPackages and grab the first package with `kind: ZarfInitConfig`

File: src/internal/packager/helm/chart.go [hunk 13]
```diff
@@ -54,6 +54,10 @@ type InstallUpgradeOptions struct {
 	Timeout time.Duration
 	// Retries for the helm install/upgrade
 	Retries int
+	// Pkg is the zarf package being installed
+	Pkg *v1alpha1.ZarfPackage
```
Comment (AustinAbro321) [inline src/internal/packager/helm/chart.go]: `.Metdata.Name` is the only field used from the package currently. Are you thinking that you might want other fields from the package in the future? Maybe a `string` or `v1alpha1.ZarfMetadata` would be more appropriate.

File: src/pkg/cluster/cluster.go [hunk 14]
```diff
@@ -154,6 +159,8 @@ type InitStateOptions struct {
 	ArtifactServer state.ArtifactServerInfo
 	// StorageClass of the k8s cluster Zarf is initializing
 	StorageClass string
+	// Pkg is the Zarf package being initialized
+	Pkg *v1alpha1.ZarfPackage
```
Comment (AustinAbro321) [inline src/pkg/cluster/cluster.go]: This is no longer needed since you're not labeling the namespace anymore

File: src/internal/packager/helm/zarf.go (hunk 12 above)
Comment (sgettys) [reply src/internal/packager/helm/zarf.go]: Can we guarantee only a single ZarfInitConfig type?

File: src/internal/packager/helm/chart.go (hunk 13 above)
Comment (sgettys) [reply src/internal/packager/helm/chart.go]: Possibly, more of a future proofing in case we want to do more package based labeling in the future. I don't have a specific use case for it now though so I don't mind narrowing the scope of that variable

Comment (sgettys) [reply src/internal/packager/helm/chart.go]: I guess we can use the package name and the namespace override label if it exists to look up the package later! So I can trim this down to be just the name

File: src/internal/packager/helm/zarf.go (hunk 12 above)
Comment (AustinAbro321) [reply src/internal/packager/helm/zarf.go]: I have seen before situations where there are multiple `ZarfInitConfig` packages deployed. The most complete thing to do would be check for a package with ZarfInitConfig and a component named `zarf-registry`. Same with the agent update, check for a package with `ZarfInitConfig` and a component named `zarf-agent`. 

Technically still possible for there to be two packages with those component names deployed, but I can't imagine a use case for it.

File: src/internal/packager/helm/post-render.go [hunk 15]
```diff
@@ -40,21 +40,22 @@ type renderer struct {
...
-	}
+func newRenderer(ctx context.Context, chart v1alpha1.ZarfChart, adoptExistingResources bool, c *cluster.Cluster, airgapMode bool, s *state.State, actionConfig *action.Configuration, variableConfig *variables.VariableConfig, pkgName string, namespaceOverride string) (*renderer, error) {
 	if actionConfig == nil {
 		return nil, fmt.Errorf("action configuration required to run post renderer")
 	}
 	if variableConfig == nil {
 		return nil, fmt.Errorf("variable configuration required to run post renderer")
 	}
+	if pkgName == "" {
+		return nil, fmt.Errorf("package required to run post renderer")
```
Comment (AustinAbro321) [inline src/internal/packager/helm/post-render.go]:
```suggestion
		return nil, fmt.Errorf("package name required to run post renderer")
```

File: src/internal/packager/helm/post-render.go [hunk 16]
```diff
@@ -178,6 +183,25 @@ func (r *renderer) editHelmResources(ctx context.Context, resources []releaseuti
 		if err := yaml.Unmarshal([]byte(resource.Content), rawData); err != nil {
 			return fmt.Errorf("failed to unmarshal manifest: %w", err)
 		}
+		// If the object is empty, it's a blank resource, so we skip it. If the package is nil, we're not in a package and we don't want to add labels.
```
Comment (AustinAbro321) [inline src/internal/packager/helm/post-render.go]:
```suggestion
		// If the object is empty, it's a blank resource, so we skip it. If the package name is empty we don't add labels.
```

File: src/pkg/cluster/injector.go [hunk 17]
```diff
@@ -36,7 +36,7 @@ import (
 )
 
 // StartInjection initializes a Zarf injection into the cluster.
-func (c *Cluster) StartInjection(ctx context.Context, tmpDir, imagesDir string, injectorSeedSrcs []string, registryNodePort int) error {
+func (c *Cluster) StartInjection(ctx context.Context, tmpDir, imagesDir string, injectorSeedSrcs []string, registryNodePort int, pkg *v1alpha1.ZarfPackage) error {
```
Comment (AustinAbro321) [inline src/pkg/cluster/injector.go]: Lets also use a string for package name here

File: src/internal/packager/helm/zarf.go [hunk 18]
```diff
@@ -179,3 +197,16 @@ func UpdateZarfAgentValues(ctx context.Context, opts InstallUpgradeOptions) erro
 	}
 	return nil
 }
+
+func findInitPackageWithComponent(pkgs []state.DeployedPackage, componentName string) string {
+	for _, pkg := range pkgs {
+		if pkg.Data.Kind == "ZarfInitConfig" {
```
Comment (AustinAbro321) [inline src/internal/packager/helm/zarf.go]:
```suggestion
		if pkg.Data.Kind == v1alpha1.ZarfInitConfig {
```

**Review by AustinAbro321 — APPROVED** [review]
LGTM, nice work

File: src/internal/packager/helm/post-render.go [hunk 19]
```diff
@@ -114,9 +117,11 @@ func (r *renderer) adoptAndUpdateNamespaces(ctx context.Context) error {
 		for _, serverNamespace := range namespaceList.Items {
 			if serverNamespace.Name == name {
 				existingNamespace = true
+				break
```
Comment (brandtkeller) [inline src/internal/packager/helm/post-render.go:120]: ++ good catch
